packages/amplify_api/lib/src/graphql/graphql_request_factory.dart

```diff
@@ -0,0 +1,353 @@
+      final fieldName = isId ? idFieldName : queryPredicate.field;
+
+      return <String, dynamic>{
+        fieldName: _queryFieldOperatorToPartialGraphQLFilter(
+            queryPredicate.queryFieldOperator)
+      };
+    }
+
+    // and, or, not
+    if (queryPredicate is QueryPredicateGroup) {
+      // .toShortString() is the same as expected graphql filter strings for all these,
+      // so no translation helper required.
+      final typeExpression = queryPredicate.type.toShortString();
+
+      // not
+      if (queryPredicate.type == QueryPredicateGroupType.not) {
+        if (queryPredicate.predicates.length == 1) {
+          return <String, dynamic>{
+            typeExpression: queryPredicateToGraphQLFilter(
+                queryPredicate.predicates[0], modelType)
+          };
+        }
+        // Public not() API only allows 1 condition but QueryPredicateGroup
+        // technically allows multiple conditions so explictly disallow multiple.
+        throw ApiException(
+            'Unable to translate not() with multiple conditions.');
+      }
+
+      // and, or
+      return <String, List<Map<String, dynamic>>>{
+        typeExpression: queryPredicate.predicates
+            .map((predicate) =>
+                queryPredicateToGraphQLFilter(predicate, modelType)!)
+            .toList()
+      };
+    }
+
+    throw ApiException(
+        'Unable to translate the QueryPredicate $queryPredicate to a GraphQL filter.');
+  }
+
+  /// Calls `toJson` on the model and removes key/value pairs that refer to
+  /// children when that field is null. The structure of provisioned AppSync `input` type,
+  /// such as `CreateBlogInput` does not include nested types, so they will get
+  /// an error from AppSync if they are included in mutations.
+  ///
+  /// When the model has a parent via a belongsTo, the id from the parent is added
+  /// as a field similar to "blogID" where the value is `post.blog.id`.
+  Map<String, dynamic> buildInputVariableForMutations(Model model) {
+    ModelSchema schema =
+        getModelSchemaByModelName(model.getInstanceType().modelName(), null);
+    final modelJson = model.toJson();
+
+    // If the model has a parent in the schema, get the ID of parent and field name.
+    String? belongsToModelName; // e.g. "blog"
+    String? belongsToKey; // e.g. "blogID"
+    String? belongsToValue; // the ID value to use from `post.blog.id`
+    final belongsToAssociation = getBelongsToFieldFromModelSchema(schema);
+    if (belongsToAssociation != null) {
+      belongsToModelName = belongsToAssociation.name;
+      belongsToKey = belongsToAssociation.association?.targetName;
+      belongsToValue = modelJson[belongsToModelName]?[idFieldName];
+    }
+
+    // Remove any relational fields or readonly.
+    Set<String> fieldsToRemove = schema.fields!.entries
+        .where((entry) =>
+            entry.value.association != null || entry.value.isReadOnly)
```

Comment:
callout: This was changed here from old branch: check for readonly fields in addition to related models.